lib/net.c

```diff
@@ -1200,12 +1201,102 @@ start_iface(struct gatekeeper_if *iface, unsigned int num_attempts_link_get)
 	return ret;
 }
 
+static void
+calculate_net_config_para(struct net_config *net_conf)
+{
+	int i;
+
+	/* The maximum number of RX queues on any interface. */
+	uint16_t gatekeeper_num_rx_queues = net_conf->back_iface_enabled ?
+		RTE_MAX(net_conf->front.num_rx_queues,
+			net_conf->back.num_rx_queues) :
+		net_conf->front.num_rx_queues;
+
+	/* The maximum number of TX queues on any interface. */
+	uint16_t gatekeeper_num_tx_queues = net_conf->back_iface_enabled ?
+		RTE_MAX(net_conf->front.num_tx_queues,
+			net_conf->back.num_tx_queues) :
+		net_conf->front.num_tx_queues;
+
+	const char *block_names[] = { "dynamic_conf" };
+
+	/*
+	 * The total number of lcores used by any functional block
+	 * for either RX or TX except for Dynamic Configuration.
+	 */
+	uint16_t gatekeeper_num_lcores = get_filtered_num_lcores(block_names,
+		sizeof(block_names) / sizeof(*block_names));
+
+	/*
+	 * The number of elements in the mbuf pool.
+	 *
+	 * The optimum size (in terms of memory usage) for a mempool is when
+	 * it is a power of two minus one.
+	 *
+	 * Need to provision enough memory for the worst case. According to
+	 * l3fwd sample application, it's essentially the number of RX
+	 * descriptors (across all queues for all ports), the number of TX
+	 * descriptors (across all queues for all ports), the number of packet
+	 * burst buffers (across all lcores for all bursts), the number of
+	 * slots in the packet buffer cache (across all lcores).
+	 */
+	net_conf->gatekeeper_num_mbuf = rte_align32pow2(
+		net_conf->gatekeeper_num_ports * gatekeeper_num_rx_queues *
+		net_conf->gatekeeper_num_rx_desc +
+		net_conf->gatekeeper_num_ports * gatekeeper_num_lcores *
+		net_conf->gatekeeper_max_pkt_burst +
+		net_conf->gatekeeper_num_ports * gatekeeper_num_tx_queues *
+		net_conf->gatekeeper_num_tx_desc +
+		gatekeeper_num_lcores * RTE_MEMPOOL_CACHE_MAX_SIZE) - 1;
+
+	/*
+	 * XXX #155 The size of the per-core object cache, i.e.,
+	 * number of struct rte_mbuf elements in the per-core object
+	 * cache. This should be analyzed or tested further to find
+	 * optimal value.
+	 *
+	 * Each core deals with at most gatekeeper_num_ports queues,
+	 * so the cache size should be at least
+	 * (number of ports * number of descriptors per queue), i.e.,
+	 * (gatekeeper_num_ports * GATEKEEPER_DESC_PER_QUEUE).
+	 *
+	 * Notice that, this argument must be lower or equal to
+	 * CONFIG_RTE_MEMPOOL_CACHE_MAX_SIZE and n / 1.5.
+	 * It is advised to choose cache_size to have
+	 * "n modulo cache_size == 0": if this is not the case,
+	 * some elements will always stay in the pool
+	 * and will never be used. Here, n is gatekeeper_num_mbuf.
+	 *
+	 * The maximum cache size can be adjusted in DPDK's .config file:
+	 * CONFIG_RTE_MEMPOOL_CACHE_MAX_SIZE.
+	 */
+	net_conf->gatekeeper_per_core_cache_size = RTE_MIN(
+		RTE_MEMPOOL_CACHE_MAX_SIZE,
+		net_conf->gatekeeper_num_mbuf / 1.5);
+	for (i = net_conf->gatekeeper_per_core_cache_size; i > 1; i--) {
```

Comment:
Replace `>` with `>=` in the test `i > 1`; otherwise, when `net_conf->gatekeeper_num_mbuf` is prime, `net_conf->gatekeeper_per_core_cache_size` won't change value.

Side note: funny enough, given that `net_conf->gatekeeper_num_mbuf` has the form `2^x - 1`, it has a higher than normal chance of being prime!